client/rest/src/main/java/org/elasticsearch/client/NodeSelector.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Selects nodes that can receive requests. Use with
+ * {@link RestClientActions#withNodeSelector withNodeSelector}.
+ */
+public interface NodeSelector {
+    /**
+     * Select the {@link Node}s to which to send requests. This may be called
```

Comment:
I would try and be even more explicit on what this methods does (feel free to make adjustments though), something like: "Receives a sorted list of nodes, in the order that the client would try them for a single request, and returns a sorted list of nodes selected out of the provided ones. Provided nodes may be alive ones, in which case returning an empty list means that the client will try and revive one dead node taken from the blacklist. When reviving nodes, this method will be called too with a list of sorted nodes as argument, where the first one is the one that the client would revive. The first node returned by the selector is the one that will be revived."